src/coap_openssl.c

```diff
@@ -492,138 +510,724 @@ int coap_dtls_context_set_psk(coap_context_t *ctx,
+
+      for (;;) {
+        if (PEM_read_bio_X509(in, &x, NULL, NULL) == NULL)
+            break;
+        if (!X509_STORE_add_cert(st, x)) {
+          while ((e = ERR_get_error()) != 0) {
+            int r = ERR_GET_REASON(e);
+            if (r != X509_R_CERT_ALREADY_IN_HASH_TABLE) {
+              coap_log(LOG_WARNING, "  %s at %s:%s\n",
+                       ERR_reason_error_string(e),
+                       ERR_lib_error_string(e),
+                       ERR_func_error_string(e));
+            }
+          }
+        }
+      }
+      BIO_free(in);
+      X509_free(x);
+    }
+    break;
+
+  case COAP_PKI_KEY_ASN1:
+    if (setup_data->key.asn1.public_cert && setup_data->key.asn1.public_cert_len > 0) {
+      if (!(SSL_CTX_use_certificate_ASN1(ctx, setup_data->key.asn1.public_cert_len,
+            setup_data->key.asn1.public_cert))) {
         coap_log(LOG_WARNING,
-      "*** coap_dtls_context_set_pki: DTLS: Unknown Private Key type %d for ASN1\n",
-                 setup_data->asn1_private_key_type);
+ "*** setup_pki: (D)TLS: %s: Unable to configure Server Certificate\n",
+                 "ASN1");
         return 0;
       }
-      if (!(SSL_CTX_use_PrivateKey_ASN1(pkey_type, context->dtls.ctx, setup_data->asn1_private_key, setup_data->asn1_private_key_len))) {
-        coap_log(LOG_WARNING, "*** coap_dtls_context_set_pki: DTLS: %s: Unable to configure Server Private Key\n", "ASN1");
+    }
+    else {
+      coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: No Server Certificate defined\n");
+      return 0;
+    }
+
+    if (setup_data->key.asn1.private_key &&
+             setup_data->key.asn1.private_key_len > 0) {
+      int pkey_type = map_key_type(setup_data->key.asn1.private_key_type);
+      if (!(SSL_CTX_use_PrivateKey_ASN1(pkey_type, ctx,
+            setup_data->key.asn1.private_key, setup_data->key.asn1.private_key_len))) {
+        coap_log(LOG_WARNING,
+ "*** setup_pki: (D)TLS: %s: Unable to configure Server Private Key\n",
+                 "ASN1");
         return 0;
       }
     }
     else {
-      coap_log(LOG_ERR, "*** coap_dtls_context_set_pki: DTLS: No Server Private Key defined\n");
+      coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: No Server Private Key defined\n");
+      return 0;
+    }
+
+    if (setup_data->key.asn1.ca_cert && setup_data->key.asn1.ca_cert_len > 0) {
+      /* Need to use a temp variable as it gets incremented*/
+      const uint8_t *p = setup_data->key.asn1.ca_cert;
+      X509* x509 = d2i_X509(NULL, &p, setup_data->key.asn1.ca_cert_len);
+      X509_STORE *st;
+      long e;
+      if (!x509 || !SSL_CTX_add_client_CA(ctx, x509)) {
+        coap_log(LOG_WARNING,
+ "*** setup_pki: (D)TLS: %s: Unable to configure client CA File\n",
+                  "ASN1");
+        X509_free(x509);
+        return 0;
+      }
+      st = SSL_CTX_get_cert_store(ctx);
+      /* Flush out existing errors */
+      while ((e = ERR_get_error()) != 0) { }
```

Comment:
For better readability, please move the closing parenthesis to the next line.

Reply:
Fixed